### PR TITLE
consensus, contract, mining, researcher, rpc, staking, gui: Implementation of MRC - baseline functionality

### LINKED ISSUE #1946: Add Balance Warning To Beacon Process
Adding a warning when someone's balance is low and unlikely to stake may help stop people who don't know about staking. I keep seeing users asking about other things that are unaware about the time it would take, so adding this would likely help. It could be an extra screen in the steps or a popup or just anything if their balance is low enough. I think that ideally this wouldn't show up if a user's balance is high or just show up as a check-mark or something unobtrusive to users with high enough balances

### LINKED ISSUE #2007: Manual Reward Claim (MRC) : BountyPot
Hi Team,

It seems there is some discussion around Manual Reward Claims feature however it is part of the roadmap but no fixed date around it. I wanted to check if it would make sense to ask people to put in a reward pot for it if they feel this is an important feature. 

For myself manual reward claim seems to be a good feature to expedite as it helps us reduce the limitation around keeping high volume of coin in the wallet all the time. How i see it is .
1. GRC is rewarded for BOINC but people keep it in wallet so that staking goes smoothly. Initial staking take a lot of time due to low GRC counts.
2. If initial GRC requirement is removed then this will simply mean that more people can be onboarded quickly. 
3. GRC pulls GRC right now. you need GRC to stake to get more GRC. Only way to earn GRC without putting money is to go pool. existing boinc users would not prefer that and for newbies that too hectic.

With all those above points in mind. I am starting this issue. For example i would be interested in putting in **2000 GRC** to the pot if MRC could be expedited. 

Note: Idea is to see if people really want this feature and are interested in putting money where their mouth is and it can act as a side incentive if the demand is high.

### PR DESCRIPTION
**Pull request by jamescowens**
This is the tracking PR for Manual Reward Claims (MRC). It is from the mrc_for_pr branch, which I will update at certain points in the mrc development, which is occurring on the mrc branch on my fork. This is a nice way for people to see the diff against the current development branch and understand what is going on. I will be adding documentation to the initial comment here to explain the design choices and implementation.

One point I will make first. I have been running at least one node on testnet with this code at various commit points to guard against regression problems against our current protocol (block v11). Commit https://github.com/gridcoin-community/Gridcoin-Research/pull/2425/commits/c290fae688f7291fbccbddd264afb7c0a6ba8b43 I did full -reindex runs against both testnet and mainnet after the most essential parts of the core code is complete and both passed. ~~I have not yet set up an isolated testnet environment to fork the chain and begin testing MRC. We are not quite ready for that. The rpc function to create an MRC transaction needs to be written (@div72 has graciously volunteered to do that) first, and I need to do a thorough static review of the block v12 consensus to ensure everything is going to work.~~ Initial testing and troubleshooting on isolated testnet fork complete. Minor issues corrected and first and second successful MRCs conducted (see below).

This PR will only have an rpc function to create the MRC. We will do a separate PR for the GUI code as this PR is big enough as it is, although the transaction categories in the GUI have been extended to properly account for MRC.

Closes #2007. Also refer to https://github.com/gridcoin-community/Gridcoin-Tasks/issues/218.

Also closes #1946 as an alternative approach.

See https://github.com/gridcoin-community/Gridcoin-Tasks/issues/250 for the poll to decide the fee split between the staker and the foundation.

# Design documentation:

- Node initiates an MRC via the createmrcrequest rpc function and/or the GUI equivalent (which will be in a separate PR). Note that the createmrcrequest will prevent a node from creating an mrc that is in the zero payment interval, or causes the number of transactions in the memory pool to exceed the MRCOutputLimit (see below). For now this can be overridden for testing purposes, but this will not be allowed on mainnet later. An optional fee can be provided above the minimum (but less than the research subsidy) to change the "queue" position of the MRC request.
- The node sends out a burn transaction (with a burn requirement of 0.01 GRC) with an attached signed MRC contract to the network (this uses the CreateMRC function).
- This MRC contract contains much the same information as a CLAIM, which includes the accrued research rewards and CBR to be paid as of the last block (chain head) before the MRC transaction time. The MRC contract is signed by the researcher's beacon key and the hash bound into the signature includes the cpid of the reseacher requesting an MRC and the block hash of the last block before the MRC was created (i.e. the head of the chain). See GetMRCHash(), TrySignMRC(), and MRCSign().
- The MRC contract transactions from all requesting nodes sit in the memory pool until a staking node binds them into a block.
- A staking node, which can be a researcher or investor, will call CreateRestOfTheBlock. This will sweep up normal transactions in priority order, and will also include MRC transactions, up to the MRCOutputLimit of ~~5 or 4~~ 10 or 9 if the Foundation MRC fee sidestake is active. MRC transactions bound to the block but whose number exceeds the MRCOutputLimit are expended (the burn goes through) without action. To minimize the possibility of this occurring, which causes the loss of the burn fee by the MRC requestor without payment, the createmrcrequest on the sending node checks the number of MRC transactions in the memory pool before creating a new one. CreateRestOfTheBlock will also check if an MRC payment has been initiated by the same CPID as the staker (which means the very node that has requested the MRC is now staking immediately after). In that situation, the MRC request will be ignored (and expended) as a normal transaction in favor of a normal research rewards stake by the researcher. This will not be very common for practical purposes. Anyone with a GRC balance high enough to stake often will NOT be using MRC as a matter of course, because it would not be rational in that instance to pay the extra fees for the MRC. People using MRC are very likely to have lower balances, which means the probability of them staking on the block right after the MRC is created by them is extremely low.
- The CreateMRCRewards, run after CreateGridcoinReward (which computes the staker's reward) will act on the MRC contract transactions included in the block to be staked and create outputs on the coinstake corresponding to the MRC contracts. This information will be recorded in an extended CLAIM contract in the m_mrc_tx_map. This map is keyed by CPID and therefore ensures only one MRC payment can exist per unique CPID in a given block.
- There is a ValidateMRC function which is wrapped by an MRC contract handler and implements the virtual Validate function to validate incoming MRC transactions. Invalid MRC transactions will cause the sending node to accumulate misbehavior (DoS) score and could lead to banning the sending node to protect against invalid MRCs. The ValidateMRC function (in unwrapped form) is also used in the GRC Validator class, which operates at the block level to validate incoming blocks as part of ConnectBlock.
- The CreateMRC function calls ComputeMRCFee to compute the fee. This function has a fee curve that looks like the following:
image
- Looking at the dark blue line, which is the mrc fees charged as a % of the MRC requesters reward: this starts at 100% during the zero payout interval (consensus parameter MRCZeroPaymentInterval), which is 14 days for mainnet and 10 minutes for testnet. While the normal createmrcrequest will refuse to form an MRC during this interval, if someone modifies a node to try and get around it, and submits an MRC, if greater than 1/2 the interval has elapsed, this will be accepted, but the MRC requester will forfeit 100% of their rewards as fees to the staker and/or foundation. If less than 1/2 the interval has elapsed, the transaction will be rejected and banscore assigned.
- After the zero payout interval has passed, the fees immediately drop to InitialMRCFeeFractionPostZeroInterval, which is 2/5 or 40%. The fees then decline as c/t where t is the interval since the last reward payment and the current MRC request. So for example at mainnet the end of the zero payment interval is 14 days, and the fees drop to 40%. At 28 days, they will be 20%, etc.
- The ComputeMRCFee function is called both when the MRC contract is created to record the fees in the MRC, AND also again by the ValidateMRC function upon receipt of a transaction by a receiving node to check that the fees were not calculated improperly by a rogue sender. It is also called by the staking node during the CreateRestOfTheBlock function. Fees that do not agree or incorrect rewards on an incoming MRC transaction will fail validation and result in misbehavior points.
- A portion of the MRC fee can be set aside by setting the Fraction FoundationSidestakeAllocation. ~~Currently this is static in the early test stage at 1/20 (5%), but it is anticipated that this will be made a dynamic setting that can be changed by an administrative protocol setting.~~ By poll ID 651a3d7cbb797ee06bd8c2b17c415223d77bb296434866ddf437a42b6d1e9d89, the foundation sidestake allocation was set to the permanent value of 4/5, i.e. 80%. A change to this value will require a poll and another mandatory.
- Likewise if the above foundation fraction is greater than zero, another function, the FoundationSidestakeAddress, will return the foundation's address. This is currently overridable by a startup parameter for testing and testnet purposes ~~, but it is anticipated that this will also be made a dynamic setting that can be changed by an administrative protocol setting~~.
- If there is a non-zero allocation to the foundation from the fees, the first MRC output on the coinstake will be the fee payment to the foundation. Essentially this is a special sidestake.
 - The remainder of the fees will go back to the staker. This is why the splitting/regular sidestake stuff has to be done AFTER the MRC payments are created, because you need to know all of the fees first.
- The accrual system has been extended to ensure it properly deals with the extended claim payments, which include MRC outputs.
- Note that the MRC payment interval and associated rewards are recorded in the MRC contracts is computed from the last payment to the last block (head of the chain) when the MRC request was created. This last block is recorded in the MRC contract and is checked by the staking node to ensure that the last block is the same as the head of the chain underneath the new block being staked. If they don't match, the MRC transaction is ignored for MRC payout. Essentially this means, in correlation with coinstakes, the MRC transactions do not survive forks/reorgs. The ending interval for MRC payouts in the coinstake is always one block behind the research payout for the staker themselves (because the staker's interval is computed on the new block being staked). MRC transactions and payouts not surviving forks/reorgs can cause MRC payments to be rolled back but this has the advantage of reducing the potentiality of double spend/double reward problems and the alignment along the block times just like the coinstake itself for the main researcher payment greatly simplifies the accrual system modifications.

### CODE REVIEW

File: src/miner.cpp [hunk 1]
```diff
@@ -1225,13 +1514,22 @@ void StakeMiner(CWallet *pwallet)
...
+
         LOCK(cs_main);
 
         g_timer.GetTimes(function + "lock cs_main", "miner");
 
         CBlockIndex* pindexPrev = pindexBest;
 
         // * Create a bare block
+
+        // This transition code is for mandatory change from V11 to v12 block format (accomodates MRC).
```
Comment (barton2526) [inline src/miner.cpp]: typo, should be "accommodates"

Comment (jamescowens) [reply src/miner.cpp]: You are annoying!! But I love you anyway! Who else is going to catch these typos! :)

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -2425,3 +2426,87 @@ UniValue getburnreport(const UniValue& params, bool fHelp)
...
+        throw runtime_error("createmrcrequest <force> <dry_run>\n"
+                            "\n"
+                            "<force> - If true, create the request even if it results "
+                            "in a reward loss or ban from the network. Defaults to false.\n"
+                            "\n"
+                            "<dry_run> - If true, calculate the reward and fee but do not "
+                            "send the contract. Defaults to false.\n"
+                            "\n"
+                            "Creates an MRC request. Requires an unlocked wallet.");
+    }
```
Comment (RoboticMind) [inline src/rpc/blockchain.cpp]: These don't appear to be required arguments, so they should follow the convention for optional arguments
```suggestion
        throw runtime_error("createmrcrequest [force [dry_run]]\n"
                            "\n"
                            "[force] - If true, create the request even if it results "
                            "in a reward loss or ban from the network. Defaults to false.\n"
                            "\n"
                            "[dry_run] - If true, calculate the reward and fee but do not "
                            "send the contract. Defaults to false.\n"
                            "\n"
                            "Creates an MRC request. Requires an unlocked wallet.");
    }
```

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -2425,3 +2426,87 @@ UniValue getburnreport(const UniValue& params, bool fHelp)
 
     return json;
 }
+
+UniValue createmrcrequest(const UniValue& params, const bool fHelp) {
+    if (fHelp || params.size() > 2) {
+        throw runtime_error("createmrcrequest <force> <dry_run>\n"
+                            "\n"
+                            "<force> - If true, create the request even if it results "
+                            "in a reward loss or ban from the network. Defaults to false.\n"
```
Comment (RoboticMind) [inline src/rpc/blockchain.cpp]: Having a force seems like it could lead to problems especially if someone meant to do a dry run but forgot an argument. What's the rationale for including this? Is this meant more for testing?

**Review by RoboticMind — COMMENTED** [review]
Looked at the RPC command and have a few comments

File: src/rpc/blockchain.cpp (hunk 3 above)
Comment (jamescowens) [reply src/rpc/blockchain.cpp]: Yes. This is not the final form for the rpc.

Comment (jamescowens) [reply src/rpc/blockchain.cpp]: The force will be removed after testing is done.

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (jamescowens) [reply src/rpc/blockchain.cpp]: I agree but I will do it on the mrc branch and flow it through.

**Comment by jamescowens** [comment]
The first successful MRC on an isolated two node testnet!
Gridcoin first successful MRC send on isolated testnet network 20220116
[first successful MRC on isolated testnet.log](https://github.com/gridcoin-community/Gridcoin-Research/files/7878349/first.successful.MRC.on.isolated.testnet.log)

**Comment by jamescowens** [comment]
Note that the characterization of the MRC fees going to the staker (in this case 100% because we are in the 14 day period), have to be recharacterized. They are showing up as PoS mined. We will have to create a separate category I think. otherwise it will be confusing.

**Comment by jamescowens** [comment]
Ok. I have implemented InitialMRCFeeFractionPostZeroInterval and MRCZeroPaymentInterval chainparams. This is to formally support different params for the mainnet and testnet chains. Note that InitialMRCFeeFractionPostZeroInterval  is set to 2/5 (40%) for both, but MRCZeroPaymentInterval is set to 14 days on mainnet, but 10 minutes on testnet to facilitate rapid testing of the zero interval and fee decay curve.

**Comment by jamescowens** [comment]
Here is another test with the above parameters in force for testnet and where the MRC is beyond the zero payment interval...
Gridcoin second successful MRC send on isolated testnet network 20220117
[second successful MRC on isolated testnet.log](https://github.com/gridcoin-community/Gridcoin-Research/files/7879465/second.successful.MRC.on.isolated.testnet.log)

**Comment by jamescowens** [comment]
I will be thinking through how to change the GUI transaction display to be more clear what is going on.

**Comment by jamescowens** [comment]
We need to decide how to extend

enum MinedType
{
    UNKNOWN = 0,
    POS = 1,
    POR = 2,
    ORPHANED = 3,
    POS_SIDE_STAKE_RCV = 4,
    POR_SIDE_STAKE_RCV = 5,
    POS_SIDE_STAKE_SEND = 6,
    POR_SIDE_STAKE_SEND = 7,
    SUPERBLOCK = 8
};

for mrcs.

This is largely for display purposes as the underlying transaction structure is already good.

I think we need MRC_RCV and MRC_SEND. The lower part of the block should always be purple because the MRC is ALWAYS associated with a research reward. The upper part is green and red, respectively for receive and send as usual.

We also may need a separate one for the foundation fee sidstake.

Let's go through an example MRC request and payment.

image

The bottom transaction is the original request. This does not have a destination, because it is a burn transaction. The debited amount is correct and good and represents the burn fee + the normal transaction fee.

The top transaction is the payment back to node 1 from the block staked by node 2.
Notice that the amoutn of

The request was initiated from node 1 (on the left), which is a node with a valid beacon and accruing research rewards to be paid, while the staking node is an investor node and on the right.

The original mrc request was as follows on node 1:

15:48:56
createmrcrequest


15:48:57
{
"txid": "0dedfc8bc1527825077278ba5189038b44b49b1cd0df6cc6ceccecec8c7a4b37",
"reward": 139.71643518,
"fee": 11.97569444
}

So you can see the reward and the fee to be taken out. Here is how node 1 displays the original request and the result.
image

The bottom transaction is the original mrc request and shows the proper debit for the burn + transaction fee. There is no destination because it is a burn.

The top transaction is the payment to node 1 from the block staked by node 2. The amount is correct: 139.71643518 - 11.97569444 = 127.74074074, which is the net payment due after taking out the MRC fees. The misleading point is the golden color of the block body. This is because the MRC was paid from a staker that is a non-researcher. This is where we need to introduce the MRC_RCV to have the bottom colored purple, regardless of whether the node staking and paying the MRC is a researcher or non-researcher, because in this case the researcher context is from the requester.

**Comment by jamescowens** [comment]
I will go through node 2's display shortly and talk about what we need to do for it too...

**Comment by jamescowens** [comment]
The display for node 2 is the following:
image

The bottom entry is the entire value of the coinstake, following the rule established long ago when side-staking was implemented, to show the entire value of the coinstake as a positive, then show the payout as negatives.

The middle entry is the MRC "sidestake" of fees to the foundation address. This only appears here as a receive because I am temporarily using an address on node 2 as the "foundation address" for staking. So this entry would not be present on mainnet in any normal person's wallet.

The top entry is the actual payment to the MRC requester.

How does the math work? The outputs minus the input is equal to 139.71643518 (MRC rewards) + 0.001 (tx fees to staker) + 10 (CBR) = 149.71743518

The fees to the staker and the foundation are taken from the MRC rewards. The foundation fraction (5%) is sent to the foundation = 0.59878472. The staker simply keeps the remainder of the fees and the MRC rewards - total fees is sent to the MRC requester. Here is a small Excel grid to show how it all fits together...
image

**Comment by jamescowens** [comment]
Ok. With the "Correct UI display/categorization of MRC payments commit applied, node 1 (the MRC requester) shows:
image
image

Node 2 (the staker) shows:
image

This is correct.

File: src/gridcoin/claim.h [hunk 4]
```diff
@@ -187,6 +187,19 @@ class Claim : public IContractPayload
     //!
     SuperblockPtr m_superblock;
 
+    //!
+    //! \brief A map of mrc transactions keyed by CPID. There can be only one MRC per CPID
+    //! in the claim.
+    //!
+    std::map<Cpid, uint256> m_mrc_tx_map;
```
Comment (denravonska) [inline src/gridcoin/claim.h:194]: Consider `unordered_map` if order doesn't matter.

File: src/gridcoin/contract/contract.cpp [hunk 5]
```diff
@@ -856,6 +861,9 @@ ContractPayload Contract::Body::ConvertFromLegacy(const ContractType type) const
             // Claims can only exist in a coinbase transaction and have no
             // legacy representation as a contract:
             assert(false && "Attempted to convert legacy claim contract.");
+        case ContractType::MRC:
+            // MRCs have no legacy representation as a contract.
+            assert(false && "Attempted to convert non-existent legacy MRC contract.");
```
Comment (denravonska) [inline src/gridcoin/contract/contract.cpp:866]: I think you want a break here or it will fall through when built with `-DNDEBUG`

Comment (div72) [reply src/gridcoin/contract/contract.cpp:866]: We don't allow builds with NDEBUG.

File: src/gridcoin/claim.h (hunk 4 above)
Comment (jamescowens) [reply src/gridcoin/claim.h:194]: Probably better, although I have to refork my testnet to test it, since it is a structural change in the claim. There can be at most 5 elements in this map, so I am not sure it really matters for a map that small.

**Comment by jamescowens** [comment]
I marked this ready for review even though some things still need tidying up.

File: src/gridcoin/claim.h (hunk 4 above)
Comment (denravonska) [reply src/gridcoin/claim.h:194]: Oh, yeah, then it doesn't matter.

File: src/gridcoin/contract/contract.cpp (hunk 5 above)
Comment (jamescowens) [reply src/gridcoin/contract/contract.cpp:866]: Technically I suppose break should be after each one, but in this particular switch, there is either a return statement or assert, so the break would never get executed.

File: src/gridcoin/claim.h (hunk 4 above)
Comment (jamescowens) [reply src/gridcoin/claim.h:194]: Do we really care that much?

Comment (jamescowens) [reply src/gridcoin/claim.h:194]: The order does not matter. You can't store output offsets (index numbers) in the claim because the claim is created BEFORE the splitting/sidestake of the main research reward, so the index of the MRC outputs can shift. We need to maintain the order of the offsets the way they are to minimize changes to other parts of the code... empty... main stake return (and splits) ... sidestakes ... (foundation mrc fee sidestake) ... mrc payments.

Comment (jamescowens) [reply src/gridcoin/claim.h:194]: This requires us to walk the map in the ValidateMRCRewards function and also walk the output vector of the coinstake when we verify the correlation of the outputs on the coinstake to the MRCs bound in the claim, but again the number of elements is extremely small on the map.

File: src/gridcoin/mrc.cpp [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+{
+}
+
+bool MRC::WellFormed() const
+{
+    if (m_version <= 0 || m_version > MRC::CURRENT_VERSION) {
+        return false;
+    }
+
+    if (m_version == 1) {
```
Comment (denravonska) [inline src/gridcoin/mrc.cpp]: Should this check exist? Setting the version to 1 bypasses the rest of the checks.

File: src/gridcoin/mrc.cpp [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+    const CpidOption cpid = m_mining_id.TryCpid();
+
+    if (!cpid) return fee;
+
+    const ResearchAccount& account = Tally::GetAccount(*cpid);
+    const int64_t last_reward_time = account.LastRewardTime();
+
+    // Get the block index of the head of the chain at the time the MRC was filled out.
+    CBlockIndex* prev_block_pindex = mapBlockIndex[m_last_block_hash];
```
Comment (denravonska) [inline src/gridcoin/mrc.cpp]: Consider using `find` and then dereference to avoid inserting data into the map on lookup failures.

**Review by denravonska — DISMISSED** [review]

File: src/gridcoin/mrc.cpp (hunk 6 above)
Comment (jamescowens) [reply src/gridcoin/mrc.cpp]: Ah no. Good catch. This should have been removed when I copied the class over from claim. :)

**Comment by jamescowens** [comment]
Ok. This is after creating a GUI category for the MRC request contract send... I created a file  src/qt/res/icons/tx_contract_mrc.svg which is a copy of the send.svg in the light theme as a temporary stand in.
image

The tooltip shows "Manual Rewards Claim Request".

File: src/gridcoin/mrc.cpp (hunk 7 above)
Comment (jamescowens) [reply src/gridcoin/mrc.cpp]: Done.

**Comment by jamescowens** [comment]
Note: There is more logging than we need in a production state. Some of the LogPrintf's will be converted to LogPrints using the MISC or VERBOSE category to cut down on the logging noise as we get further through testing on testnet.

**Comment by jamescowens** [comment]
There is missing serialization needed in CBlockIndex. I will be adding this tonight.

File: src/gridcoin/tally.cpp [hunk 8]
```diff
@@ -290,6 +292,69 @@ class ResearcherTally
...
+        pindex = pindex->pprev;
+
+        while (pindex)
+        {
+            // If it is a stake with this cpid, then stop
+            if (pindex->ResearchSubsidy() > 0 && pindex->GetMiningId() == cpid) break;
+
+            // If it is an MRC with this cpid and non-zero reward, then stop
+            for (const auto& mrc_researcher : pindex->m_mrc_researchers) {
+                if (mrc_researcher->m_cpid == cpid && mrc_researcher->m_research_subsidy > 0) break;
```
Comment (div72) [inline src/gridcoin/tally.cpp]:
```suggestion
                if (mrc_researcher->m_cpid == cpid && mrc_researcher->m_research_subsidy > 0) {
                    pindex = pindex->pprev;
                    break;
                }
```
The true last reward would be the block before the MRC claim. See the comments below for why this is necessary.

File: src/gridcoin/tally.cpp [hunk 9]
```diff
@@ -322,17 +387,52 @@ class ResearcherTally
             account.m_total_magnitude -= pindex->Magnitude();
         }
 
-        pindex = pindex->pprev;
+        pindex = FindLastRewardBlock(cpid, pindex);
```
Comment (div72) [inline src/gridcoin/tally.cpp:421]: If the researcher here had used MRC to claim their rewards and this stake was rewinded, FindLastRewardBlock would point to one block further than the true reward height.

File: src/gridcoin/tally.cpp [hunk 10]
```diff
@@ -322,17 +387,52 @@ class ResearcherTally
...
+
+            account.m_total_research_subsidy -= mrc_researcher->m_research_subsidy;
+
+            if (mrc_researcher->m_magnitude > 0) {
+                account.m_accuracy--;
+                account.m_total_magnitude -= pindex->pprev->Magnitude();
+            }
+
+            // This search is based on the contents of pindex and so takes and outputs pindex as the argument even for MRCs.
+            pindex = FindLastRewardBlock(cpid, pindex);
```
Comment (div72) [inline src/gridcoin/tally.cpp]: Reverse ditto. Somebody who staked before the MRC would have their last_block_ptr set to a block behind the true block height.

File: src/main.cpp [hunk 11]
```diff
@@ -954,16 +1001,40 @@ class ClaimValidator
...
+            //If the FoundationSideStakeAllocation is active, then we have to check the outputs on the coinstake returning
+            // to the staker and ensure that they minus the input do not exceed
+            // research_owed + out_stake_owed + m_fees + staker_fees_owed.
+            if (FoundationSideStakeAllocation().isNonZero()) {
+
+                const CTransaction& coinstake = m_block.vtx[1];
+                CAmount total_owed_to_staker = -m_stake_value_in;
+
+                for (const auto& output : coinstake.vout) {
+                    if (output.scriptPubKey == coinstake.vout[1].scriptPubKey) {
```
Comment (div72) [inline src/main.cpp]: Can't someone game this by side-staking the amount to another address? Checking the amount sent to the foundation feels safer to me here.

File: src/gridcoin/mrc.cpp [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool MRC::HasResearchReward() const
+{
+    return m_mining_id.Which() == MiningId::Kind::CPID;
```
Comment (div72) [inline src/gridcoin/mrc.cpp]: Is this method used?

File: src/gridcoin/mrc.cpp [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+
+    if (!m_mining_id.Valid()) {
+        return false;
+    }
+
+    if (m_client_version.empty()) {
+        return false;
+    }
+
+    if (m_mining_id.Which() == MiningId::Kind::CPID) {
```
Comment (div72) [inline src/gridcoin/mrc.cpp]: Investors and invalid mining ids should not be able to make MRC requests imo.

File: src/main.cpp [hunk 14]
```diff
@@ -1224,6 +1335,162 @@ class ClaimValidator
...
+                            if (contract.m_type != GRC::ContractType::MRC) continue;
+
+                            GRC::MRC mrc = contract.CopyPayloadAs<GRC::MRC>();
+
+                            if (const GRC::CpidOption cpid = mrc.m_mining_id.TryCpid()) {
+                                CBlockIndex* mrc_index = mapBlockIndex[mrc.m_last_block_hash];
+
+                                const GRC::BeaconOption beacon = GRC::GetBeaconRegistry().TryActive(*cpid, mrc_index->nTime);
+
+                                if (beacon) {
```
Comment (div72) [inline src/main.cpp:1545]: TrySignMRC will prevent expired beacons to send, should people with expired beacons be able to make MRC requests for the last time to claim their rewards?

File: src/rpc/blockchain.cpp [hunk 15]
```diff
@@ -2425,3 +2476,105 @@ UniValue getburnreport(const UniValue& params, bool fHelp)
...
+
+        std::tie(wtx, error) = GRC::SendContract(GRC::MakeContract<GRC::MRC>(GRC::ContractAction::ADD, std::move(mrc)));
+        if (!error.empty()) {
+            throw runtime_error(error);
+        }
+        resp.pushKV("txid", wtx.GetHash().ToString());
+    }
+
+    resp.pushKV("reward", (double)reward / COIN);
+    resp.pushKV("fee", (double)fee / COIN);
```
Comment (div72) [inline src/rpc/blockchain.cpp]:
```suggestion
    resp.pushKV("mrc", MRCToJson(mrc));
```
self-nit:

File: src/gridcoin/mrc.cpp (hunk 12 above)
Comment (jamescowens) [reply src/gridcoin/mrc.cpp]: Nope. We can get rid of it. you can't actually submit an MRC in an investor context. This was just carried over when I copied the Claim class.

File: src/main.cpp (hunk 14 above)
Comment (jamescowens) [reply src/main.cpp:1545]: I don't think so. The accrual will bridge an expired beacon if someone advertises and validates a new beacon using gthe same CPID. The key will be different and they will have to send a new MRC request, but I think the strict approach is safer.

File: src/rpc/blockchain.cpp (hunk 15 above)
Comment (jamescowens) [reply src/rpc/blockchain.cpp]: Yep. Very good idea.

File: src/gridcoin/mrc.cpp (hunk 13 above)
Comment (jamescowens) [reply src/gridcoin/mrc.cpp:81]: Yes, and that is what the validation code does at 81-83 does, in combination with 73-75.

File: src/main.cpp (hunk 11 above)
Comment (jamescowens) [reply src/main.cpp]: Hmm... I will ponder this tonight.

File: src/gridcoin/tally.cpp (hunk 8 above)
Comment (jamescowens) [reply src/gridcoin/tally.cpp]: I think you are right. This is tricky stuff, because the accrual account last block/first block has to record the block at which the accrual ends but the actual block that the mrc is recorded is the block with the claim/coinstake of the mrc. In the case of normal rewards this is coincident, but with mrc, the accrual block context for the reward is always one block behind. I fixed the other parts of the tally to deal with this, but I think missed it in this one place. I think this commit fixes all of the below too.

File: src/gridcoin/tally.cpp (hunk 9 above)
Comment (jamescowens) [reply src/gridcoin/tally.cpp:421]: I think this is fixed by the one line commit above.

File: src/gridcoin/tally.cpp (hunk 10 above)
Comment (jamescowens) [reply src/gridcoin/tally.cpp]: Ditto.

File: src/gridcoin/tally.cpp [hunk 16]
```diff
@@ -322,17 +390,52 @@ class ResearcherTally
...
+
+            account.m_last_block_ptr = pindex->pprev;
```
Comment (div72) [inline src/gridcoin/tally.cpp]:
```suggestion
            account.m_last_block_ptr = pindex;
```
Missed this, the other two comments were just to show to clarify the point.

Comment (jamescowens) [reply src/gridcoin/tally.cpp]: Wait a second. Isn't that just a shell game? whether you put the step to pprev here or in the search, doesn't it result in the same thing?

Comment (jamescowens) [reply src/gridcoin/tally.cpp]: The issue you were describing earlier is precisely why I had the m_last_block_ptr set to pindex->pprev here. One of us is confused, and I am not sure whom at this point.

Comment (jamescowens) [reply src/gridcoin/tally.cpp]: Well... I think we played a shell game here, but applying your suggested commit above fixes it.

Comment (div72) [reply src/gridcoin/tally.cpp]: > whether you put the step to pprev here or in the search, doesn't it result in the same thing?

Not exactly.

Let's assume a chain consisting of Stake(0) -> NOP(1) -> MRC(2) -> NOP(3) -> Stake(4) -> NOP(5) -> MRC(6) for a researcher. When the chain rollbacks one block(6), ForgetMRCRewardBlock is triggered and calls FindLastRewardBlock. This method would return the block number 4 in this case. If the ForgetMRCRewardBlock used pprev here, it would mistakenly refer to 3 instead of 4 because of the pprev.

Going a step further and rolling back two more blocks(fifth block is not relevant), rolling back number 4 would trigger ForgetRewardBlock. The function itself would call FindLastRewardBlock which would return 2. Since MRC payments are based off on the last block, the actual height should've been 1 but it's missing a pprev.

Moving the logic to FindLastRewardBlock ensures that, in the first case the same number is returned. ForgetMRCRewardBlock does not have to do a pprev, so the height is correctly 4. In the second case, the function would return 1 correctly and no modification would be needed to do for ForgetRewardBlock.

Comment (div72) [reply src/gridcoin/tally.cpp]: I'll add a unit test for this, so you can check.

File: src/main.cpp (hunk 11 above)
Comment (jamescowens) [reply src/main.cpp]: Yep. I think there is definitely a hole. I am recoding this and will post as soon as I do a smoke test.

**Comment by jamescowens** [comment]
Here is a screenshot of a successful two node MRC request (i.e. each simultaneously request) with a two MRC payout from the third node.
image

Note the first test of this caught a misplaced break, which was fixed in https://github.com/gridcoin-community/Gridcoin-Research/pull/2425/commits/3d6ed1552ed1fe5afbcd88354a65daaa977a8e3d!

**Comment by jamescowens** [comment]
The unit tests are failing again after I made tweaks to the MRC validation. I will take a look and fix.

**Comment by jamescowens** [comment]
Fixed.

**Comment by jamescowens** [comment]
I redid the sync from zero test on both testnet and mainnet (this is v11 only) to recheck for regression issues with pre-v12 (MRC) sync. Both passed with flying colors.

**Comment by jamescowens** [comment]
I am doing another sync from zero on my isolated testnet branch (fork #3) of one of the nodes to ensure that the transition to v12 goes well. If this passes I am merging this PR into the development branch. The MRC code will be inactive on testnet and is backwards compatible with earlier code for block versions v11 and below. The remaining items to actually activate MRC on testnet will be done in a subsequent PR.

**Comment by jamescowens** [comment]
The sync from zero through v12 transition in the isolated fork went flawlessly. Merging this PR into development.
